s3torchconnector/tst/e2e/test_e2e_s3_lightning_checkpoint.py

```diff
@@ -0,0 +1,217 @@
+    dataset = WikiText2(data_dir=Path(f"/tmp/data/{nonce}"))
+    dataloader = DataLoader(dataset, num_workers=3)
+    model = LightningTransformer(vocab_size=dataset.vocab_size)
+    trainer = L.Trainer(fast_dev_run=2)
+    trainer.fit(model=model, train_dataloaders=dataloader)
+    checkpoint_name = "lightning_module_training_checkpoint.pt"
+    s3_uri = f"{checkpoint_directory.s3_uri}{checkpoint_name}"
+    trainer.save_checkpoint(s3_uri)
+    s3_lightning_checkpoint = S3LightningCheckpoint(region=checkpoint_directory.region)
+    loaded_checkpoint = s3_lightning_checkpoint.load_checkpoint(s3_uri)
+    _verify_equal_state_dict(model.state_dict(), loaded_checkpoint["state_dict"])
+
+
+def test_compatibility_with_trainer_plugins(checkpoint_directory):
+    nonce = random.randrange(2**64)
+    dataset = WikiText2(data_dir=Path(f"/tmp/data/{nonce}"))
+    dataloader = DataLoader(dataset, num_workers=3)
+    model = LightningTransformer(vocab_size=dataset.vocab_size)
+    s3_lightning_checkpoint = S3LightningCheckpoint(region=checkpoint_directory.region)
+    trainer = L.Trainer(
+        default_root_dir=checkpoint_directory.s3_uri,
+        plugins=[s3_lightning_checkpoint],
+        max_epochs=1,
+        max_steps=3,
+    )
+    trainer.fit(model, dataloader)
+    checkpoint_key = "lightning_logs/version_0/checkpoints/epoch=0-step=3.ckpt"
+    checkpoint_s3_uri = f"{checkpoint_directory.s3_uri}{checkpoint_key}"
+    loaded_checkpoint = s3_lightning_checkpoint.load_checkpoint(checkpoint_s3_uri)
+    _verify_equal_state_dict(model.state_dict(), loaded_checkpoint["state_dict"])
+
+    new_model = LightningTransformer.load_from_checkpoint(
+        checkpoint_s3_uri, vocab_size=dataset.vocab_size
+    )
+    _verify_equal_state_dict(model.state_dict(), new_model.state_dict())
+
+
+def test_compatibility_with_checkpoint_callback(checkpoint_directory):
+    nonce = random.randrange(2**64)
+    dataset = WikiText2(data_dir=Path(f"/tmp/data/{nonce}"))
+    dataloader = DataLoader(dataset, num_workers=3)
+
+    model = LightningTransformer(vocab_size=dataset.vocab_size)
+    s3_lightning_checkpoint = S3LightningCheckpoint(checkpoint_directory.region)
+
+    checkpoint_callback = ModelCheckpoint(
+        dirpath=checkpoint_directory.s3_uri,
+        save_top_k=1,
+        every_n_epochs=1,
+        filename="checkpoint-{epoch:02d}-{step:02d}",
+        enable_version_counter=True,
+    )
+    trainer = L.Trainer(
+        plugins=[s3_lightning_checkpoint],
+        callbacks=[checkpoint_callback],
+        min_epochs=4,
+        max_epochs=5,
+        max_steps=3,
+    )
+
+    trainer.fit(model, dataloader)
+    expected_checkpoint_name = "checkpoint-epoch=00-step=03.ckpt"
+    # Workaround to verify there is only one checkpoint file. Client should NOT be accessed this way.
+    bucket, prefix = parse_s3_uri(checkpoint_directory.s3_uri)
+    list_result = list(s3_lightning_checkpoint._client.list_objects(bucket, prefix))
```

Comment:
Is there any reason why we are going to use approach that we explicitly call a "bad" one ourself?

Reply:
Good catch, forgot the client is available there 😄